dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,335 @@
+        }
+        else {
+            winrt::VisualStateManager::GoToState(*this, c_previousPageButtonHiddenVisualState, false);
+            winrt::VisualStateManager::GoToState(*this, c_previousPageButtonDisabledVisualState, false);
+        }
+
+        if (selectedPageIndex != numberOfPages - 1) {
+            if (m_isPointerOver) {
+                winrt::VisualStateManager::GoToState(*this, c_nextPageButtonVisibleVisualState, false);
+            }
+            winrt::VisualStateManager::GoToState(*this, c_nextPageButtonEnabledVisualState, false);
+        }
+        else {
+            winrt::VisualStateManager::GoToState(*this, c_nextPageButtonHiddenVisualState, false);
+            winrt::VisualStateManager::GoToState(*this, c_nextPageButtonDisabledVisualState, false);
+        }
+    }
+}
+
+void PipsControl::OnElementPrepared(winrt::ItemsRepeater sender, winrt::ItemsRepeaterElementPreparedEventArgs args)
+{
+    if (const auto pip = args.Element().try_as<winrt::Button>()) {
+        if (unbox_value<int>(pip.Tag()) - 1 != SelectedPageIndex()) {
+            winrt::VisualStateManager::GoToState(pip, L"Unselected", true);
+        }
+
+        const auto buttonClickedFunc = [this](auto const& sender, auto const& args) {
+            if (const auto button = sender.try_as<winrt::Button>())
+            {
+                SelectedPageIndex(winrt::unbox_value<int>(button.Tag()) - 1);
+            }
+        };
+        pip.Click(buttonClickedFunc);
+    }
+}
+
+void PipsControl::ScrollToCenterOfViewport(winrt::UIElement sender)
+{
+    winrt::BringIntoViewOptions options;
+    options.VerticalAlignmentRatio(0.5);
+    options.AnimationDesired(true);
+    sender.StartBringIntoView(options);
+}
+
+
+void PipsControl::OnMaxDisplayedPagesChanged(const int oldValue) {
+    m_lastMaxDisplayedPages = oldValue;
+    auto const numberOfPages = NumberOfPages();
+    if (m_lastMaxDisplayedPages != numberOfPages) {
+        UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
+    }
+}
+
+void PipsControl::OnNumberOfPagesChanged(const int oldValue)
+{
+    m_lastNumberOfPagesCount = oldValue;
+    const int numberOfPages = NumberOfPages();
+    if (SelectedPageIndex() > numberOfPages - 1 && numberOfPages > -1) {
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
+    UpdateNavigationButtonVisualStates();
+}
+
+void PipsControl::OnSelectedPageIndexChange(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    // Special case is NumberOfPages being 0, in that case, don't verify upperbound restrictions
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > 0)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
+    else {
```

Comment:
nit:newline to be consistent with the formatting.

Reply:
Microsoft c++ code bases all put { on new lines

---
In reply to: [520979619](https://github.com/microsoft/microsoft-ui-xaml/pull/3592#discussion_r520979619) [](ancestors = 520979619)